crates/astria-merkle/src/lib.rs

```diff
@@ -0,0 +1,718 @@
+    /// string, i.e. `MTH({}) = Sha256()`.
+    #[must_use]
+    pub fn root(&self) -> [u8; 32] {
+        if self.is_empty() {
+            Sha256::digest(b"").into()
+        } else {
+            self.get_node(complete_root(self.len()))
+        }
+    }
+
+    /// Creates a new, empty merkle tree.
+    ///
+    /// # Examples
+    /// ```
+    /// # use astria_merkle::Tree;
+    /// let tree = Tree::new();
+    /// assert_eq!(0, tree.len());
+    /// ```
+    #[must_use]
+    pub fn new() -> Self {
+        Self {
+            nodes: Vec::new(),
+        }
+    }
+
+    /// Returns the number of nodes in the merkle tree.
+    ///
+    /// # Examples
+    /// ```
+    /// # use astria_merkle::Tree;
+    /// let mut tree = Tree::new();
+    /// tree.push(&[1; 32]);
+    /// assert_eq!(1, tree.len());
+    /// // Pushing a second leaf will also insert a parent to hold their combined hash.
+    /// tree.push(&[2; 32]);
+    /// assert_eq!(3, tree.len());
+    /// ```
+    #[must_use]
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.nodes.len() / 32
+    }
+
+    #[must_use]
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.nodes.is_empty()
+    }
+
+    /// Build a leaf hash ad-hoc by writing its bytes into a [`LeafBuilder`].
+    ///
+    /// The leaf is added to the tree when [`LeafBuilder`] is dropped.
+    ///
+    /// If a leaf itself is build up from several byte slices, this method allows
+    /// adding a leaf hash to the merkle tree without having to allocate the leaf first.
+    ///
+    /// # Examples
+    /// ```
+    /// # use astria_merkle::Tree;
+    /// let mut tree_1 = Tree::new();
+    /// // The `LeafBuilder` returned by `Tree::build_leaf` is never assigned, so it
+    /// // goes out of scope at the end of the statement, updating the tree on drop.
+    /// tree_1.build_leaf().write(b"hello").write(b"world");
+    ///
+    /// let mut leaf = Vec::new();
+    /// leaf.extend_from_slice(b"hello");
+    /// leaf.extend_from_slice(b"world");
+    /// let mut tree_2 = Tree::new();
+    /// tree_2.push(&leaf);
+    ///
+    /// assert_eq!(tree_1.root(), tree_2.root());
+    /// ```
+    pub fn build_leaf(&mut self) -> LeafBuilder<'_> {
+        let mut hasher = Sha256::new();
+        hasher.update([0x00_u8]);
+        LeafBuilder {
+            tree: self,
+            hasher: Some(hasher),
+        }
+    }
+
+    /// Pushes a new leaf into the tree.
+    pub fn push(&mut self, leaf: &[u8]) {
+        self.build_leaf().write(leaf);
+    }
+
+    /// Constructs a Merkle tree from an iterator yielding byte slices.
+    ///
+    /// This is a utility function to loop over an iterator and pushing
+    /// each item into the tree.
+    pub fn from_leaves<I, B>(iter: I) -> Self
+    where
+        I: IntoIterator<Item = B>,
+        B: AsRef<[u8]>,
+    {
+        let mut tree = Self::new();
+        for item in iter {
+            tree.push(item.as_ref());
+        }
+        tree
+    }
```

Comment:
since `push` updates the tree every time it's called, it would be nice to have a `from_leaves` function that doesn't call `push`, but just constructs the tree directly from all the leaves. this way, we can do 1 bigger allocation at once since we know the final tree size, and don't have to hash intermediates unnecessarily. thoughts?

Reply:
I would love to have that, but to really benefit from this we would need to walk the tree level by level and bottom to top, and only update branches at each level.

But the math comes out a bit annoying and is IMO out of scope for this PR. I would want to spend some time thinking about how to do that.

A thought: according the indexing scheme the trailing set bits indicate the depth of the subtree. So the root for a tree of depth `D` has a binary representation of `011...1` with `D-1` ones. This extends to every subtree. This means we should in principle be able to walk the tree by "roots-of-subtrees" by fixing the number of trailing 1s for each level.